lock/src/test/java/org/infinispan/lock/impl/lock/RequestExpirationSchedulerTest.java

```diff
@@ -0,0 +1,121 @@
+package org.infinispan.lock.impl.lock;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.testng.AssertJUnit.assertEquals;
+import static org.testng.AssertJUnit.assertFalse;
+import static org.testng.AssertJUnit.assertTrue;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.mockito.ArgumentCaptor;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "clusteredLock.RequestExpirationSchedulerTest")
+public class RequestExpirationSchedulerTest {
+
+   private RequestExpirationScheduler expirationScheduler;
+   private ScheduledExecutorService scheduledExecutorService = mock(ScheduledExecutorService.class);
+   private ScheduledFuture scheduledFutureMock;
+   private CompletableFuture<Boolean> booleanCompletableFuture;
+
+   @BeforeMethod
+   public void createRequestExpirationScheduler() {
+      expirationScheduler = new RequestExpirationScheduler(scheduledExecutorService);
+      scheduledFutureMock = mock(ScheduledFuture.class);
+      booleanCompletableFuture = new CompletableFuture();
+      when(scheduledExecutorService.schedule(any(Runnable.class), anyLong(), any(TimeUnit.class))).thenReturn(scheduledFutureMock);
+   }
+
+   @Test
+   public void testScheduleForCompletionAndRunScheduling() throws Exception {
+
+      ArgumentCaptor<Runnable> argument = ArgumentCaptor.forClass(Runnable.class);
+
+      expirationScheduler.scheduleForCompletion("123", booleanCompletableFuture, 10, TimeUnit.MILLISECONDS);
+      assertEquals(1, expirationScheduler.countScheduledRequests());
+
+      verify(scheduledExecutorService).schedule(argument.capture(), eq(10L), eq(TimeUnit.MILLISECONDS));
+
+      // We capture the argument so we can run as if the scheduler worked and assert
+      Runnable runnable = argument.getValue();
+      runnable.run();
+
+      assertTrue(booleanCompletableFuture.isDone());
+      assertFalse(booleanCompletableFuture.get());
+      assertEquals(0, expirationScheduler.countScheduledRequests());
+   }
+
+   @Test
+   public void testScheduleForCompletionAddsASingleRequestById() throws Exception {
+
+      expirationScheduler.scheduleForCompletion("123", booleanCompletableFuture, 100, TimeUnit.SECONDS);
```

Comment:
This also seems like a bug that if we schedule 2 requests on the same id (like my comment on the impl). I would normally expect some kind of error condition as there is no way to then timeout the second or third requests as well.

Reply:
I agree, that it does not fail and it should throw an exception